// @gathertown/gather-game-client/src/DefaultSubscriptions.ts
export const defaultSubscriptions = {
  playerJoins: true,
  playerSetsAffiliation: true,
  playerSetsStatus: true,
  playerSpotlights: true,
  playerRings: true,
  playerInteracts: true,
  playerSetsImagePointer: true,
  playerChats: true,
  playerSetsOutfitString: true,
  playerMoves: true,
  playerSetsName: true,
  playerGhosts: true,
  playerSetsEmoteV2: true,
  playerSetsWorkCondition: true,
  playerActivelySpeaks: true,
  playerSetsTextStatus: true,
  playerSetsEmojiStatus: true,
  playerEntersWhisperV2: true,
  playerLeavesWhisper: true,
  playerSetsIsSignedIn: true,
  playerBlocks: true,
  spaceRegistersCommand: true,
  playerSendsCommand: true,
  mapSetDimensions: true,
  mapSetCollisions: true,
  mapSetBackgroundImagePath: true,
  mapSetForegroundImagePath: true,
  mapSetSprites: true,
  mapSetSpaces: true,
  mapSetSpawns: true,
  mapSetPortals: true,
  mapSetAnnouncer: true,
  mapSetAssets: true,
  mapSetObjects: true,
  mapSetName: true,
  mapSetMuteOnEntry: true,
  mapSetUseDrawnBG: true,
  mapSetWalls: true,
  mapSetFloors: true,
  mapSetAreas: true,
  mapDeleteObject: true,
  mapSetSpawn: true,
  mapSetMiniMapImagePath: true,
  mapSetEnabledChats: true,
  mapSetDescription: true,
  mapSetDecoration: true,
  mapSetTutorialTasks: true,
  mapSetScript: true,
  mapSetDesks: true,
  mapCommitsChanges: true,
  mapMoveObject: true,
  spaceOverwrites: true,
  spaceIsClosed: true,
  spaceSetsIdMapping: true,
  playerEntersPortal: true,
  playerExits: true,
  info: true,
  warn: true,
  error: true,
  playerSetsLastActive: true,
  playerShootsConfetti: true,
  playerSetsEventStatus: true,
  playerSetsInConversation: true,
  playerSetsCurrentDesk: true,
  playerSetsCurrentArea: true,
  playerTriggersItem: true,
  cookieFound: true,
  playerSetsVehicleId: true,
  playerSetsIsAlone: true,
  spacePlaysSound: true,
  playerSetsIsMobile: true,
  setScreenPointerServer: true,
  playerSetsFocusModeEndTime: true,
  playerSetsItemString: true,
  spaceSetsSpaceMembers: true,
  customEvent: true,
  playerUpdatesFocusModeStatus: true,
  playerNotifies: true,
  playerSetsFollowTarget: true,
  playerRequestsToLead: true,
  playerSetsManualVideoSrc: true,
  playerSetsIsNpc: true,
  playerSetsSubtitle: true,
  speakerUpdatesSession: true,
  playerEditsChatMessage: true,
  playerHighFives: true,
  playerSetsSpeedModifier: true,
  fxShakeObject: true,
  fxShakeCamera: true,
  playerUpdatesInventory: true,
  spaceUpdatesItems: true,
};
