// @gathertown/gather-game-common/dist/src/public/gameMap.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
exports.__esModule = true;
exports.WallFloorDB = exports.wallFloorConvertToDB = exports.wallFloorConvertFromDB = exports.WallBlock = exports.verifyMapIsComplete = exports.TileType = exports.Tile = exports.PresetTutorialGroupSetIds = exports.MAP_MAX_DIMENSION = exports.interactionOpensModal = exports.Interaction = exports.generateEmptyDeskCoordsMap = exports.FloorBlock = exports.findMissingMapObjectFields = exports.EnabledChat = exports.convertWireTutorialTasksToTutorialTasks = exports.convertWireObjectToMapObjectDB = exports.convertWireObjectToMapObject = exports.convertWireAreasToDBAreas = exports.convertWireAreasToAreas = exports.convertTutorialTasksToWireTutorialTasks = exports.convertStringToPresetTutorialGroupSetId = exports.convertStringToEnabledChat = exports.convertMapObjectToWireObject = exports.convertDesksToDBDesks = exports.convertDBDesksToDesks = exports.convertDBDesksCoordsToCoordsMap = exports.convertCoordsToBBoxes = exports.convertAreaCoordsToCoordsMap = exports.areaToCoords = exports.AreaCategory = void 0;
var GameMap_1 = require("../GameMap");
__createBinding(exports, GameMap_1, "AreaCategory");
__createBinding(exports, GameMap_1, "areaToCoords");
__createBinding(exports, GameMap_1, "convertAreaCoordsToCoordsMap");
__createBinding(exports, GameMap_1, "convertCoordsToBBoxes");
__createBinding(exports, GameMap_1, "convertDBDesksCoordsToCoordsMap");
__createBinding(exports, GameMap_1, "convertDBDesksToDesks");
__createBinding(exports, GameMap_1, "convertDesksToDBDesks");
__createBinding(exports, GameMap_1, "convertMapObjectToWireObject");
__createBinding(exports, GameMap_1, "convertStringToEnabledChat");
__createBinding(exports, GameMap_1, "convertStringToPresetTutorialGroupSetId");
__createBinding(exports, GameMap_1, "convertTutorialTasksToWireTutorialTasks");
__createBinding(exports, GameMap_1, "convertWireAreasToAreas");
__createBinding(exports, GameMap_1, "convertWireAreasToDBAreas");
__createBinding(exports, GameMap_1, "convertWireObjectToMapObject");
__createBinding(exports, GameMap_1, "convertWireObjectToMapObjectDB");
__createBinding(exports, GameMap_1, "convertWireTutorialTasksToTutorialTasks");
__createBinding(exports, GameMap_1, "EnabledChat");
__createBinding(exports, GameMap_1, "findMissingMapObjectFields");
__createBinding(exports, GameMap_1, "FloorBlock");
__createBinding(exports, GameMap_1, "generateEmptyDeskCoordsMap");
__createBinding(exports, GameMap_1, "Interaction");
__createBinding(exports, GameMap_1, "interactionOpensModal");
__createBinding(exports, GameMap_1, "MAP_MAX_DIMENSION");
__createBinding(exports, GameMap_1, "PresetTutorialGroupSetIds");
__createBinding(exports, GameMap_1, "Tile");
__createBinding(exports, GameMap_1, "TileType");
__createBinding(exports, GameMap_1, "verifyMapIsComplete");
__createBinding(exports, GameMap_1, "WallBlock");
__createBinding(exports, GameMap_1, "wallFloorConvertFromDB");
__createBinding(exports, GameMap_1, "wallFloorConvertToDB");
__createBinding(exports, GameMap_1, "WallFloorDB");
