// @gathertown/gather-game-common/dist/src/public/player.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
exports.__esModule = true;
exports.INVENTORY_MAX_ROW = exports.INVENTORY_MAX_COLUMN = exports.whisperIdToColor = exports.TEMPORARY_EMOTES = exports.SpriteDirection = exports.positionAfterMove = exports.oppositeMoveDirection = exports.nextSpriteDirection = exports.MoveDirection = exports.keyPressedToEmote = exports.getMoveDirFromSpriteDir = exports.getAllEmotes = exports.generateDefaultPlayer = exports.EMOTE_SHORTCUTS = exports.EMOTE_EMOJIS = exports.Emote = exports.directionToFacePlayer = exports.DEFAULT_TEMP_EMOTES = exports.CONSTANT_PERMANANT_EMOTES = void 0;
var Player_1 = require("../Player");
__createBinding(exports, Player_1, "CONSTANT_PERMANANT_EMOTES");
__createBinding(exports, Player_1, "DEFAULT_TEMP_EMOTES");
__createBinding(exports, Player_1, "directionToFacePlayer");
__createBinding(exports, Player_1, "Emote");
__createBinding(exports, Player_1, "EMOTE_EMOJIS");
__createBinding(exports, Player_1, "EMOTE_SHORTCUTS");
__createBinding(exports, Player_1, "generateDefaultPlayer");
__createBinding(exports, Player_1, "getAllEmotes");
__createBinding(exports, Player_1, "getMoveDirFromSpriteDir");
__createBinding(exports, Player_1, "keyPressedToEmote");
__createBinding(exports, Player_1, "MoveDirection");
__createBinding(exports, Player_1, "nextSpriteDirection");
__createBinding(exports, Player_1, "oppositeMoveDirection");
__createBinding(exports, Player_1, "positionAfterMove");
__createBinding(exports, Player_1, "SpriteDirection");
__createBinding(exports, Player_1, "TEMPORARY_EMOTES");
__createBinding(exports, Player_1, "whisperIdToColor");
__createBinding(exports, Player_1, "INVENTORY_MAX_COLUMN");
__createBinding(exports, Player_1, "INVENTORY_MAX_ROW");
