// @gathertown/gather-game-common/dist/src/public/position.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
exports.__esModule = true;
exports.serializePositionCoordinates = exports.serializePosition = exports.positionToIndex = exports.manhattanDistance = exports.isPosWithinRect = exports.isPosEqual = exports.indexToPosition = exports.getPosFromBoundingBox = exports.getCoordsFromCoordsMap = exports.getBoundingBox = exports.euclideanDistance = exports.deserializePosition = exports.coordinatesToIndex = exports.calcCentroid = void 0;
var Position_1 = require("../Position");
__createBinding(exports, Position_1, "calcCentroid");
__createBinding(exports, Position_1, "coordinatesToIndex");
__createBinding(exports, Position_1, "deserializePosition");
__createBinding(exports, Position_1, "euclideanDistance");
__createBinding(exports, Position_1, "getBoundingBox");
__createBinding(exports, Position_1, "getCoordsFromCoordsMap");
__createBinding(exports, Position_1, "getPosFromBoundingBox");
__createBinding(exports, Position_1, "indexToPosition");
__createBinding(exports, Position_1, "isPosEqual");
__createBinding(exports, Position_1, "isPosWithinRect");
__createBinding(exports, Position_1, "manhattanDistance");
__createBinding(exports, Position_1, "positionToIndex");
__createBinding(exports, Position_1, "serializePosition");
__createBinding(exports, Position_1, "serializePositionCoordinates");
