// @gathertown/gather-game-common/dist/src/public/remoteWork.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
exports.__esModule = true;
exports.TRIVIA_MAP_ID = exports.TRIVIA_COLORS = exports.TriviaGamePhase = exports.QUESTION_PACKS_INFO = exports.QUESTION_PACKS = exports.DEFAULT_QUESTION_VALUE = exports.SocialRoomInfo = exports.SocialRoom = exports.ROOM_OF_REQUIREMENTS_EXTENSION = exports.ROOM_DOOR_OBJECT_ID = exports.ROOM_CONTROLLER_OBJECT_ID = exports.isRorStandaloneObject = exports.isRorObject = void 0;
var RemoteWorkRoomOfRequirements_1 = require("../RemoteWorkRoomOfRequirements");
__createBinding(exports, RemoteWorkRoomOfRequirements_1, "isRorObject");
__createBinding(exports, RemoteWorkRoomOfRequirements_1, "isRorStandaloneObject");
__createBinding(exports, RemoteWorkRoomOfRequirements_1, "ROOM_CONTROLLER_OBJECT_ID");
__createBinding(exports, RemoteWorkRoomOfRequirements_1, "ROOM_DOOR_OBJECT_ID");
__createBinding(exports, RemoteWorkRoomOfRequirements_1, "ROOM_OF_REQUIREMENTS_EXTENSION");
__createBinding(exports, RemoteWorkRoomOfRequirements_1, "SocialRoom");
__createBinding(exports, RemoteWorkRoomOfRequirements_1, "SocialRoomInfo");
var RemoteWorkTrivia_1 = require("../RemoteWorkTrivia");
__createBinding(exports, RemoteWorkTrivia_1, "DEFAULT_QUESTION_VALUE");
__createBinding(exports, RemoteWorkTrivia_1, "QUESTION_PACKS");
__createBinding(exports, RemoteWorkTrivia_1, "QUESTION_PACKS_INFO");
__createBinding(exports, RemoteWorkTrivia_1, "TriviaGamePhase");
__createBinding(exports, RemoteWorkTrivia_1, "TRIVIA_COLORS");
__createBinding(exports, RemoteWorkTrivia_1, "TRIVIA_MAP_ID");
__exportStar(require("../RemoteWorkFishing"), exports);
