// @gathertown/gather-game-common/dist/src/RemoteWorkTrivia.js
"use strict";
var _a;
exports.__esModule = true;
exports.DEFAULT_QUESTION_VALUE = exports.QUESTION_PACKS_INFO = exports.QUESTION_PACKS = exports.TRIVIA_COLORS = exports.TriviaGamePhase = exports.TRIVIA_MAP_ID = void 0;
exports.TRIVIA_MAP_ID = "rw-trivia";
var TriviaGamePhase;
(function (TriviaGamePhase) {
    TriviaGamePhase["UNINITIALIZED"] = "UNINITIALIZED";
    TriviaGamePhase["JOIN"] = "JOIN";
    TriviaGamePhase["SHOW_QUESTION"] = "SHOW_QUESTION";
    TriviaGamePhase["SHOW_ANSWER"] = "SHOW_ANSWER";
    TriviaGamePhase["ENDED"] = "ENDED";
})(TriviaGamePhase = exports.TriviaGamePhase || (exports.TriviaGamePhase = {}));
exports.TRIVIA_COLORS = new Set([
    "#FF9DDE",
    "#FBD54E",
    "#64E9D1",
    "#AEE964",
    "#A966FF",
    "#FC7079",
    "#FF9960",
    "#EA36AC",
    "#B67800",
    "#2D62EC",
    "#37C214",
    "#7230FF",
    "#D42E2E",
    "#E85110",
]);
var QUESTION_PACKS;
(function (QUESTION_PACKS) {
    QUESTION_PACKS["MIXED_250"] = "mixed_250";
    QUESTION_PACKS["GENERAL_200"] = "general_200";
})(QUESTION_PACKS = exports.QUESTION_PACKS || (exports.QUESTION_PACKS = {}));
exports.QUESTION_PACKS_INFO = (_a = {},
    _a[QUESTION_PACKS.MIXED_250] = {
        id: "mixed_250",
        fileName: "Mixed_250.csv",
        numQuestions: 250,
        categories: [
            "Movies",
            "Music",
            "News Events",
            "Pop Culture",
            "Sports",
            "TV",
            "Video Games",
            "Business",
            "Computers & Technology",
            "Literature",
        ],
        maxDifficulty: 3,
        displayName: "Mixed Categories"
    },
    _a[QUESTION_PACKS.GENERAL_200] = {
        id: "general_200",
        fileName: "General_200.csv",
        numQuestions: 200,
        categories: ["General Trivia"],
        maxDifficulty: 3,
        displayName: "General Trivia"
    },
    _a);
exports.DEFAULT_QUESTION_VALUE = 100;
