// @gathertown/gather-game-common/src/RemoteWorkRoomOfRequirements.ts
import { MapObject } from "./GameMap";
import { Portal } from "./generated_DO_NOT_TOUCH/events";

export const ROOM_OF_REQUIREMENTS_EXTENSION = "RemoteWorkRoomOfRequirements";

export const ROOM_CONTROLLER_OBJECT_ID =
  "CyberpunkKioskWall - ja17fdCaXI-xbC9Imc0Rm_754e70d9-0b7b-4fc2-bebe-e2c1207c95d2";
export const ROOM_DOOR_OBJECT_ID = "gBm9cPiE9ORWWatgyeC9_dedd43b1-6d34-4511-a92e-c5e019b8617a";

export const isRorObject = (id: string) =>
  id === ROOM_CONTROLLER_OBJECT_ID || id.startsWith("RoomOfRequirements");
export const isRorStandaloneObject = (id: string) => id.startsWith("RoomOfRequirementsStandalone");

export enum SocialRoom {
  GoKarts = "Go Karts",
  SpeedGathering = "Speed Gathering",
  TetrisTournament = "Tetris Tournament",
  Trivia = "Trivia",
  Poker = "Poker",
}

export type RoomOfRequirementsSetRoom = {
  rorObj: { id: string; x: number; y: number };
  currentRoom?: SocialRoom;
  selectedRoom: SocialRoom;
  myPlayerMap: string;
  myPlayerId?: string;
};

export interface TemplatePortal {
  x: number;
  y: number;
  targetOffsetX?: number;
  targetOffsetY?: number;
}

export const SocialRoomInfo: {
  [roomKey in SocialRoom]: {
    caption: string;
    imageURL: string;
    guideURL: string;
    portal: Omit<Portal, "x" | "y">; // portals from main to ror template
    templatePortals: TemplatePortal[]; // return portals from ror template to main
    roomDisplayObject: { wall?: Partial<MapObject>; standalone: Partial<MapObject> }; // wall is old
  };
} = {
  [SocialRoom.GoKarts]: {
    caption: "Host a race with your team. Watch out for stray cones to avoid set-backs!",
    imageURL: "/images/remoteWorkOnboarding/gokarts-preview.gif",
    guideURL:
      "https://support.gather.town/help/gather-grand-prix?utm_source=template&utm_medium=app&utm_campaign=experiences",
    portal: {
      targetMap: "rw-racetrack",
      targetX: 3,
      targetY: 85,
    },
    templatePortals: [
      { x: 2, y: 87 },
      { x: 3, y: 87, targetOffsetX: 1 },
    ],
    roomDisplayObject: {
      wall: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town-dev.appspot.com/uploads/m4S94FU2KzRo9quK/DSgL3qeRMCaWVbG1LTP3y9",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town-dev.appspot.com/uploads/m4S94FU2KzRo9quK/mVyx2iMwCHn4DH0zMdNzVk",
      },
      standalone: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/Nd42XW8xix7dgSiQ/Z9Om7qrQuWlSGs8jaaGtJE",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/Nd42XW8xix7dgSiQ/bWM3HIqI1LjMaVVTIjlZEc",
      },
    },
  },
  [SocialRoom.SpeedGathering]: {
    caption: "Get to know your teammates through some rapid-fire questions!",
    imageURL: "/images/remoteWorkOnboarding/speed-gathering-preview.gif",
    guideURL:
      "https://support.gather.town/help/speed-gathering?utm_source=template&utm_medium=app&utm_campaign=experiences",
    portal: {
      targetMap: "rw-speed-gathering",
      targetX: 5,
      targetY: 22,
    },
    templatePortals: [
      { x: 4, y: 24 },
      { x: 5, y: 24, targetOffsetX: 1 },
    ],
    roomDisplayObject: {
      wall: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/7gByoh17OL7iPxbP/CO4bXwwkwN8IojnFpPxYKF",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/7gByoh17OL7iPxbP/OXjpOmiVlZs7PjPoR60cAe",
      },
      standalone: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/Nd42XW8xix7dgSiQ/wYFNXd9Ih0wvoeRlkLTEH3",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/Nd42XW8xix7dgSiQ/dAFiPWozE4psQAYprUsctW",
      },
    },
  },
  [SocialRoom.TetrisTournament]: {
    caption: "Battle in your office's tetris tournament and face off for the title!",
    imageURL: "/images/remoteWorkOnboarding/tetris-preview.gif",
    guideURL:
      "https://support.gather.town/help/tetris-tourney?utm_source=template&utm_medium=app&utm_campaign=experiences",
    portal: {
      targetMap: "rw-tetris-tournament",
      targetX: 3,
      targetY: 35,
    },
    templatePortals: [
      { x: 2, y: 37 },
      { x: 3, y: 37, targetOffsetX: 1 },
    ],
    roomDisplayObject: {
      wall: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town-dev.appspot.com/uploads/m4S94FU2KzRo9quK/dZ1Onm5XhUmlZKUnphge5c",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town-dev.appspot.com/uploads/m4S94FU2KzRo9quK/cgqrarZSiY3KRgefvCWAhs",
      },
      standalone: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/Nd42XW8xix7dgSiQ/OfZ6sexRpqQQk6zRWVpxxf",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/Nd42XW8xix7dgSiQ/En1OK1M7I6deVqj8SMVs08",
      },
    },
  },
  [SocialRoom.Trivia]: {
    caption: "Pit your wits against your friends in the ultimate trivia battle!",
    imageURL: "/images/remoteWorkOnboarding/trivia-preview.gif",
    guideURL:
      "https://support.gather.town/help/trivia-pub?utm_source=template&utm_medium=app&utm_campaign=experiences",
    portal: {
      targetMap: "rw-trivia",
      targetX: 13,
      targetY: 17,
    },
    templatePortals: [
      { x: 13, y: 18 },
      { x: 14, y: 18, targetOffsetX: 1 },
    ],
    roomDisplayObject: {
      wall: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/AW5hIzlPQ4PVDGCKSjMQYG",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/OUcfsxAB7fkkASb5tNODth",
      },
      standalone: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/AW5hIzlPQ4PVDGCKSjMQYG",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/OUcfsxAB7fkkASb5tNODth",
      },
    },
  },
  [SocialRoom.Poker]: {
    caption: "Bet, check, or fold? Call your teammates' bluff in the poker room!",
    imageURL: "/images/remoteWorkOnboarding/poker-preview.gif",
    guideURL:
      "https://support.gather.town/help/poker-night?utm_source=template&utm_medium=app&utm_campaign=experiences",
    portal: {
      targetMap: "rw-poker",
      targetX: 16,
      targetY: 33,
    },
    templatePortals: [
      { x: 16, y: 34 },
      { x: 17, y: 34, targetOffsetX: 1 },
    ],
    roomDisplayObject: {
      wall: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/GjYnIUWWpGReStmRVMDDaG",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/T8tyxjy4O0orSKtInoebMS",
      },
      standalone: {
        normal:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/GjYnIUWWpGReStmRVMDDaG",
        highlighted:
          "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/T8tyxjy4O0orSKtInoebMS",
      },
    },
  },
};

// Future links
// Metaverse Cantina: https://support.gather.town/help/metaverse-cantina?utm_source=template&utm_medium=app&utm_campaign=experiences
//    normal: https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/BR5jtndXWvMdlshNdxHjtf
//    highlighted: https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/gEjZqny018p3tNJu/rakpJxQGfiC6Ynpket8rqu
// Taking Sides: https://support.gather.town/help/taking-sides?utm_source=template&utm_medium=app&utm_campaign=experiences
// Movie Night: https://support.gather.town/help/movie-night?utm_source=template&utm_medium=app&utm_campaign=experiences
